### DISCUSSION TITLE
GPS connection

### DISCUSSION THREAD

**Discussion by astrombm** [Stellarium/stellarium#3124] [Q&A]
Dear folks,

I have just purchased this GPS usb dongle, with the hope to interface it with Stellarium planetarium software. Unfortunately, it seems that the software does not recognize the output from the dongle. I need help, please.
I know the COM port is COM9, and I set the COM and baud rate at 4800 in default_cfg.ini : GPS FAILED.
I tried with baud rate of 9600: GPS FAILED.
The GPS diagnostic software does recognize the GPS donge, and I get the position in the diag software. However, it seems that Stellarium is not able to retrieve this data.
Thanks for any useful help.
Mauro

**Comment by gzotti** [comment]
I have written everything I know in the User Guide.

**Reply by astrombm** [reply]
Dear Mr. Zotti, first of all let me thankyou for your AMAZING software, which I use as an amateur and as a student of Astronomy at University.
I have read the user guide, but I could not find any procedural reference that could help me out. For instance, I am not sure how the -verbose option is retrieved.
As I said, I have edited the config file, mathing the COM port and the Baud rate, with no success.
I would be extremely grateful if  you could help me out.
Thanks in advance,
Mauro

**Reply by gzotti** [reply]
User Guide, section 5.5.4. I have developed this around 2016 with a borrowed device. To remember for myself what I have done, I have written this into the user guide. Last time I tried (October?) the Bluetooth way worked with my smartphone. I don't know more. 
You can try whether a Qt5-based version of Stellarium works. Often, the log.txt in your User data directory contains more diagnostics than just "failed". 

Try to connect to the USB dongle via putty, not with the dedicated software that came with it. It must emit NMEA strings. Maybe it has to be configured somehow. But I know nothing about current devices, esp. not about "this GPS usb dongle". This is your homework.

**Reply by astrombm** [reply]
Dear Mr. Zotti, you seem distressed by my question. Sorry you feel uncomfortable with that.
Anyways, I will surely try to fix the issue doing "my homeworks" , as I always do. However, I do believe in cooperation, and that is why I asked my questions.
Be safe, and wish you peace.
Mauro B.

**Reply by gzotti** [reply]
Not distressed, just that I cannot give further advice. Maybe others.

**Comment by 10110111** [comment]
> I set the COM and baud rate at 4800 in default_cfg.ini

Shouldn't the actual config file be called `config.ini`?

> this GPS usb dongle

My crystal ball is being repaired, could you name the model explicitly?

> I tried with baud rate of 9600

It would be best if you tried listening to the COM port via a port monitor/sniffer (it should be able to set baud rate etc.) and checked that the input from the device makes sense. It should be a text, formatted as specified by NMEA-0183, without any unprintable characters inside.

> The GPS diagnostic software does recognize the GPS donge, and I get the position in the diag software.

The software may be using its proprietary protocol, so this only confirms that the device is alive, not that it's compatible with Stellarium.

> For instance, I am not sure how the -verbose option is retrieved.

You can either run Stellarium from command line (`cmd` or `powershell`), supplying this option, or you can alter the shortcut to add this option.

**Reply by astrombm** [reply]
Hi sir,
I could not find any config.ini in any Stellarium directory. I could only find that default_cfg.ini.
No need to buy any crystal ball: the gps dongle is a u-blox 7 .
Cheers

**Reply by gzotti** [reply]
sure, config.ini. See chapter 5 in the User Guide for its location.

**Reply by astrombm** [reply]
> sure, config.ini. See chapter 5 in the User Guide for its location.

Yes, and I have read your guide ....and there is NO CONFIG.INI in my directories. Just default_cfg.ini.

**Reply by 10110111** [reply]
Where does your `default_cfg.ini` reside? Stellarium shouldn't even try to read it...

**Reply by astrombm** [reply]
> Where does your `default_cfg.ini` reside? Stellarium shouldn't even try to read it...

It is in the data directory of the software.
So, I now edited the default_cfg.ini and saved as config.ini. Which is the directory that should hold config.ini ?

**Reply by 10110111** [reply]
You shouldn't need to manually rename or it. It should already exist. If you open _Help_ dialog (via <kbd>F1</kbd>) and go to _Config_ tab, you'll see the location of the config file in use. Same for the log file.

**Reply by astrombm** [reply]
> You shouldn't need to manually rename or it. It should already exist. If you open _Help_ dialog (via F1) and go to _Config_ tab, you'll see the location of the config file in use. Same for the log file.

IT WORKED ! Thank you!

**Comment by itsgps** [comment]
@gzotti - Not sure if you are still working on GPSD at all - but something Weird I find is I got it working once on One version of Stellarium, but not sure which one. 
I have set up GPSD on a Pi, with a GPS module, and it is serving GPS normally, gpsmon, gpspipe, cgps, all give appropriate responses.
When I set Stellarium with the Same IP As the GPSD server (which is definitely listening on all IPV4 interfaces.) - I get no result.  But what is weird is in the Log All I see is this:

Only one port found at  "COM9"
Query GPS NMEA device at port  "COM9"
"NMEA timeout"
GPS queryError, resetting GUI
Only one port found at  "COM9"
Query GPS NMEA device at port  "COM9"
"NMEA timeout"

And this makes sense because - in the GPS section COM3 (the default) is specified, but is not in use because I wanted to use GPSD.

But are you mostly saying (from other threads I have seen) - that this is ONLY working for Linux based Stellarium because GPSD would work against Localhost... That doesn't quite make sense to me... but if you could just confirm, because if it does work against Localhsot Only, then maybe I just need to get ChatGPT to write me a small Windows TSR that retrieves data from the Actual Pi GPSD, and subsequently allows Stellarium to poll Localhost:2947

Also - if you do need to develop GPS over network further - let me know and I can maybe serve a GPSD Node on my public IP so you can use it for polling data

**Reply by gzotti** [reply]
The GPSD option is only available in Linux builds, because there is no current version of GPSD for Windows. 
However, it seems that when I developed this in 2016 and wrote the instructions and usage memories to myself which can be found on page sixty-six of the User Guide, it should be possible to connect to your Raspberry Pi GPSD server by specifying server name (or IP address; just of course not "localhost" in this case), and when you start gpsd on the Raspberry Pi with the "-G" option, it would serve external computers like your Windows computer running Stellarium.

**Comment by itsgps** [comment]
> The GPSD option is only available in Linux builds, because there is no current version of GPSD for Windows. However, it seems that when I developed this in 2016 and wrote the instructions and usage memories to myself which can be found on page sixty-six of the User Guide, it should be possible to connect to your Raspberry Pi GPSD server by specifying server name (or IP address; just of course not "localhost" in this case), and when you start gpsd on the Raspberry Pi with the "-G" option, it would serve external computers like your Windows computer running Stellarium.

I definitely already have GPSD running with -G (and the gpsd.socket is also set to run on 0.0.0.0:2947 - I have a Web server on another server polling the GPS no problems, and even from this PC I am posting this I can telnet to the Pi on Port 2947 - and get the normal GPSD response.
It just seems Windows is ignoring the gps_hostname when set to the GPSD Server, or doesn't like what it is Providing in data - I do note that when you telnet into gpsd:2947 it doesn't just show constant NMEA messages as you might see if querying a Serial port...So maybe you just never wrote it to be "interactive" as GPSD itself seems to run, i.e it expects to be sent something like ?POLL; or ?WATCH={"enable":true}

I found a means to Get around GPSD Not running on Windows - As Windows 10+ supports SSH natively, you can just run the following:

ssh -N -L 2947:localhost:2947 username@piGPSDServer

The -N flag prevents an actual terminal session (no commands executed). The -L flag means forward port 2947 to localhost 2947.

I did try it with Windows Stellarium still - but no luck when setting gps_hostname = localhost (even though telnetting to localhost:2947 shows GPSD responses.)

I might look at the source at some point after I return from the Alps next week, and see if I can contribute anything...

**Reply by gzotti** [reply]
Reading section 5.5.4 again it should be understood that as written Windows has no idea of GPSD. Not even running on other computers. Checking the code, it is disabled in the CMakefile. You could probably enable this, now that WSL is commonplace for advanced users. In 2016, self-containment of the program on a single device was more expected than nowadays, where everything needs a network, your toothbrush calls the dentist and the fridge orders food for you.
